### PR TITLE
sBTC-mini Stacking Pool Quick Fixes

### PR DESCRIPTION
**Pull request by setzeus**
Minor/Medium changes to address [562](https://github.com/Trust-Machines/stacks-sbtc/issues/562), [547](https://github.com/Trust-Machines/stacks-sbtc/issues/547), & [548](https://github.com/Trust-Machines/stacks-sbtc/issues/548). 

Main change is the addition of a private helper (mass-delegate-stack-extend ...) that extends stacking to either the burn address (in the form of penalize-helper) or to the newly voted-upon sBTC-wallet (in the form of a consensus vote).

Since the private function runs in different places, the response handling is inconsistent & could probably be "prettier" if any review has a suggestion for line 729.

### CODE REVIEW

**Comment by setzeus** [comment]
@FriendsFerdinand neither of these updates changes the minimal unit tests found in sbtc-stacking-pool_test.clar. We're still working on how to best POX-3 & other contracts testing 2.0 in another branch.

Fixes here are mainly to address the minor/medium issues flagged in linked issues. Minor one from Mike was removing a dummy/test function I left in a WIP. 

The two medium issues flagged by Friedger had to do with me missing "delegate-stack-extend" calls in both voting & penalizing. This PR addresses both of those issues by using a private (mass-delegate-stack-extend ...) found in line 738; this is called from either (vote-for-threshold-wallet-candidate ...) once the next peg-wallet has found voting consensus, or from one of the four (penalize-helper ...) functions once a penalty is successfully submitted.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 1]
```diff
@@ -729,14 +726,23 @@
             rewards-disbursed: bool
         }
     ))
-    (match (as-contract (contract-call? .pox-3 stack-aggregation-commit-indexed pox-burn-address next-cycle))
+    (try! (match (fold mass-delegate-stack-extend (get stackers penalized-pool) (ok {stacker: tx-sender, unlock-burn-height: u0, pox-addr: pox-burn-address}))
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:729]: Why not just return the result of the fold call?

**Comment by jo-tm** [comment]
We are following the Diagram in the Main Doc, STACKING:
image

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 2]
```diff
@@ -540,19 +522,33 @@
...
-                            false
+
+                    ;; 70% consensus reached, ready to set next cycle threshold-wallet
+                    ;; Extend delegate all signers in pool & attempt to aggregate-commit-index
+                    (begin 
+                        
+                        ;; Assert that extend-and-commit-bool is true (both delegate-stack-extend & aggregate-commit-indexed succeeded)
+                        (asserts! (match (fold mass-delegate-stack-extend next-pool-stackers (ok {stacker: tx-sender, unlock-burn-height: u0, pox-addr: pox-addr}))
+                            passed-result
+                                (match (as-contract (contract-call? .pox-3 stack-aggregation-commit-indexed pox-addr next-cycle))
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:533]: `stack-aggregation-commit-indexed` should be called the first time the 70% condition passed, thereafter `stack-aggregation-commit-increase` needs to be called with the index from the first call.

See https://github.com/friedger/clarity-stacking-pools/blob/main/contracts/pox-pool-self-service.clar#L175

**Comment by jo-tm** [comment]
SBTC MINTING (PEG-IN):
image

**Comment by friedger** [comment]
To be able to handle the penalty we must to not lock the stackers during the voting window.
Instead, we should call the extend with either burn address or peg-wallet during the penalty window.

The penalty window must start before the pox prepare phase. Maybe 200 blocks before the start of the cycle.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 2 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:533]: Thinking through this I don't believe we need `stack-aggregation-commit-increase`?

1. The pool of stackers registered in the previous cycle
2. Whether they vote for the winning wallet, they vote for the losing wallet, or they _abstain_ from the vote, they should still be part of the extend-delegate + aggregate commit with the winning wallet
3. Once the winning wallet reaches consensus (> 70% stacking weight), we don't care for additional votes
4. Only registered signers can vote & that list of registered signers is fixed (aka no one can register & vote in the same cycle)

Therefore once we mass-delegate-extend & aggregate-commit we should never need to aggregate-commit-increase? At least within the voting logic*

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 3]
```diff
@@ -729,14 +726,23 @@
...
+            (ok (ok u1))
         err-result
-            false
+            err-mass-delegate-stack-extend
+    ))
+)
+
+;; mass delegate stack extender helper for either penalizing or concluded vote
+(define-private (mass-delegate-stack-extend (stacker principal) (pox-return (response {stacker: principal, unlock-burn-height: uint, pox-addr: { version: (buff 1), hashbytes: (buff 32) }} uint))) 
+    (let
```
[the remark this answers is not in the document]
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:734]: Didn't get picked up but took care of this [in latest commit](https://github.com/Trust-Machines/stacks-sbtc/pull/578/commits/0df2d20e75f3e8c9d350ec981b4ae9ad4f9f7a9c) with a simple asserts: 

;; Assert previous pox-return was (ok ...)
`(asserts! (is-ok pox-return) err-mass-delegate-stack-extend)`

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 1 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: Severely overcomplicated myself, appreciate the suggestion.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 2 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:533]: Ok, make sense. If voting is not required for signers.

**Comment by friedger** [comment]
Blocked by https://github.com/Trust-Machines/stacks-sbtc/issues/582

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 4]
```diff
@@ -63,6 +63,8 @@
 (define-constant err-not-protocol-caller (err u35))
 (define-constant err-threshold-percent-out-of-range (err u36))
 (define-constant err-threshold-to-scriptpubkey (err u37))
+(define-constant err-mass-delegate-stack-extend (err u38))
+(define-constant err-wallet-consensus-reached-execution (err u39))
```
Comment (MarvinJanssen) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:67]: Just an FYI for everyone. We need to formally lay down all protocol error codes. I will bring it up the next meeting.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 5]
```diff
@@ -497,11 +477,12 @@
             next-pool-signer 
             {vote: (some pox-addr)}
         ))
+        
 
         ;; Check whether map-entry for candidate-wallet already exists
         (if (is-none current-candidate-status)
```
Comment (MarvinJanssen) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:483]: This entire function is pretty big with a lot of nesting. Is it possible to split it up into a few functions for readability?

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 6]
```diff
@@ -540,19 +522,33 @@
...
+                            err-result
+                                false
+                        ) err-wallet-consensus-reached-execution)
+
+                        true
                     )
+
                     ;; 70% consensus not reached, continue
-                    false
+                    true
```
Comment (MarvinJanssen) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:551]: This nested structure is pretty hard to follow. (Especially the many else/match err branches.) Can it be refactored to a series of `asserts!`?

**Review by MarvinJanssen — CHANGES_REQUESTED** [review]
Looks good, comments are mostly readability questions.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 5 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: Refactored with series of asserts, more in line with what you were thinking?

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 6 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: Yeap, updated whenever you get a chance to approve.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 7]
```diff
@@ -497,68 +478,66 @@
...
-            )
 
-            ;; Candidate exists, update relevant maps: votes-per-cycle, signer, & check whether this wallet is over threshold
-            (let
+            )
+            ;; Existing candidate path
+            (let 
                 (
-                    (unwrapped-candidate (unwrap! current-candidate-status err-unwrapping-candidate))
+                    (unwrapped-candidate (unwrap-panic current-candidate-status))
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:504]: current-candiate-status is-none, isn't it?
As checked here: https://github.com/Trust-Machines/stacks-sbtc/pull/578/files#diff-9e54e6dfb041bc4c76781bc43adfdf0ce7f1b980d41f9207780022f52d558cf8R487

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 8]
```diff
@@ -497,68 +478,66 @@
...
-                            ))
-                        err-result
-                            false
-                    )
-                    ;; 70% consensus not reached, continue
-                    false
-                )
-
+                                    next-pool
+                                    {last-aggregation: (some block-height), reward-index: (some (unwrap-panic (as-contract (contract-call? .pox-3 stack-aggregation-commit-indexed pox-addr next-cycle))))}
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:530]: I would put the result of the contract call into a separate line/variable.
unwrap-panic seems wrong as stack-aggregation-commit can return an error.

**Review by friedger — COMMENTED** [review]
Use of unwrap-panic is not clear to me.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 7 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:504]: Negative, I think? The (let ...) runs when one of the (and ...) conditions fail. I check (is-none current-candidate=status) is true as the first condition.

If that returns as false, aka current-candidate-status *does exist* then that triggers the (let ...) branch. Since we know current-candidate-status is a (some ...) we can safely unwrap-panic.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 9]
```diff
@@ -497,68 +478,66 @@
...
+                        aggregate-commit-index: none,
+                        votes-in-ustx: (get amount next-pool-signer),
+                        num-signer: u1,
                 })
-                
                 ;; Update pool map by appending wallet-candidate to list of candidates
                 (map-set pool next-cycle (merge 
                     next-pool
-                    {threshold-wallet-candidates: (unwrap! (as-max-len? (append (get threshold-wallet-candidates next-pool) pox-addr) u100) err-too-many-candidates)}
+                    {threshold-wallet-candidates: (unwrap-panic (as-max-len? (append (get threshold-wallet-candidates next-pool) pox-addr) u100))}
```
[the remark this answers is not in the document]
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:497]: Copy. Changing back. I think I'm now swinging the other way on using unwrap-panic instead of overusing (unwrap! ...). 

Any chance you could summarize the conditions on when you find it appropriate to specifically unwrap-panic?

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 8 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:543]: Changed back to a match, will only use panic sparingly / as we define.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 9 above)
[the remark this answers is not in the document]
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: I would use unwrap-panic only when I am sure that panic won't happen.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 7 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:505]: OK, the git diff was not clear to me.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 4 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:67]: Added an issue: https://github.com/Trust-Machines/stacks-sbtc/issues/590

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 10]
```diff
@@ -497,68 +479,71 @@
...
+                            )
+                        )
+                    (ok u1))
+                    
+                    (ok u2)
             )
         )
 
-        (ok true)
+        (ok u3)
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:546]: Codes should be constants with some meaningful names.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 11]
```diff
@@ -729,14 +715,22 @@
...
-        err-result
-            false
+    (fold mass-delegate-stack-extend (get stackers penalized-pool) (ok {stacker: tx-sender, unlock-burn-height: u0, pox-addr: pox-burn-address}))
+)
+
+;; mass delegate stack extender helper for either penalizing or concluded vote
+(define-private (mass-delegate-stack-extend (stacker principal) (pox-return (response {stacker: principal, unlock-burn-height: uint, pox-addr: { version: (buff 1), hashbytes: (buff 32) }} uint))) 
+    (let
+        (
+            (extend-delegate-response (unwrap! (as-contract (contract-call? .pox-3 delegate-stack-extend stacker pox-burn-address u1)) err-mass-delegate-stack-extend))
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:725]: Idea for optimization: use `as-contract` around `(fold mass-delegate-stack-extend` To be verified whether it makes a difference.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 12]
```diff
@@ -497,68 +479,71 @@
...
+                            (>= (/ (* new-candidate-votes u1000) next-pool-total-stacked) (var-get threshold-consensus))
+                            
+                            (match (as-contract (contract-call? .pox-3 stack-aggregation-commit-indexed pox-addr next-cycle))
+                                ok-index
+                                    (map-set pool next-cycle (merge 
+                                        next-pool
+                                        {last-aggregation: (some block-height), reward-index: (some ok-index)}
+                                    ))
+                                err-result
+                                    ;; Chore - anything additional in this branch?
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:536]: We could emit a print event

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 13]
```diff
@@ -497,68 +479,71 @@
...
-        (if (is-none current-candidate-status)
-
-            ;; Candidate doesn't exist, update relevant maps: votes-per-cycle & pool
-            (begin 
-                ;; Update votes-per-cycle map with first vote for this wallet-candidate
-                (map-set votes-per-cycle {cycle: next-cycle, wallet-candidate: pox-addr} {
-                    aggregate-commit-index: none,
-                    votes-in-ustx: (get amount next-pool-signer),
-                    num-signer: u1,
+        (asserts!
```
Comment (friedger) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:484]: I find this control flow hard to understand. Suggestion:

```
(if (is-none current-candidate-status) (add-wallet-candiate ....) (update-wallet-candidate...))
(maybe-aggregate-commit...)
```

Can a whale trigger the aggregate commit? If not

```
(if (is-none current-candidate-status) 
   ;; handle new candidate
   (add-wallet-candiate ....) 
   ;; handle existing candidate
   (begin
     (update-wallet-candidate...)
     (maybe-aggregate-commit...)))
```

**Review by friedger — CHANGES_REQUESTED** [review]
Needs rebase.

I have an issue with the control flow of signer-pre-register

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 12 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:536]: Great idea, doing so.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 10 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:546]: Copy. Err responses are clear to me how to identify with uint, but how/what would you use to identify an (ok ...) response? 

Same thing as error? Just a number saved as more meaningful string constant?

Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:546]: Yes, like `(ok result-committed)`

Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:546]: Or better `ok-committed`

Comment (MarvinJanssen) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:546]: Given our design pattern for error constants, I think the latter makes the most sense.

```clarity
(define-constant ok-committed (ok u3))
```

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 11 above)
Comment (MarvinJanssen) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:725]: It should make a difference.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 12 above)
Comment (MarvinJanssen) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:536]: Good idea, we do not have a comprehensive spec for custom events yet. (Something I will bring up later.)

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 13 above)
Comment (MarvinJanssen) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:484]: Both if branches can be abstracted into two private functions.

Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:484]: Removed the (if ...) that was previously there in exchange for that series of asserts! Should I change back?

Re: Hoisting Out Private Funcs
Great discussion point here. My personal rule-of-thumb is it's harder to hop around private-functions than it is to read a single well-commented / well-labeled function. 

If I need a series of functions for more than one place, of course I'll hoist out into a private function to no repeat, otherwise though, if it's only callable by one other function I'll leave it as-is. 

Thoughts here? Good opportunity to capture & document in the new docs Marvin introduced.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 11 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:734]: Wasn't picked up but updated with the (as-contract ...) wrapped around the (fold ...).

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 13 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:489]: I like to read in particular public functions without going into unnecessary details. I think that is also true for the user of a contract.
In a second go I would verify whether the private functions indeed implement what the name suggests.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 14]
```diff
@@ -523,25 +523,21 @@
...
-                    (ok u1))
+                    ok-vote-existing-candidate-lost)
                     
-                    (ok u2)
+                    ok-vote-existing-candidate-won
             )
         )
 
-        (ok u3)
+        ok-voted
```
Comment (MarvinJanssen) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar]: These nested `asserts!` are still a little tricky to follow. It is not immediately clear what these short-circuit branches belong to.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 11 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:719]: The `as-contract` should be removed here. @setzeus

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 13 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:482]: Let's move this discussion to the document or somewhere else and let's move this PR forward.

I suggest to merge this PR and open a new issue for better readability.

**Review by friedger — CHANGES_REQUESTED** [review]
The last thing I would like to see is the removal of `as-contract` in `mass-delegate-stack-extend` and tests are failing.

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 14 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: We can improve readability in a new PR

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 12 above)
Comment (friedger) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: events would be out of scope for this PR

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 11 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:724]: Not picked up but now removed, all conversations should be resolved / ready.

**Review by friedger — APPROVED** [review]
Let's merge this and optimize in a separate PR

**Comment by friedger** [comment]
@MarvinJanssen Do you have time to review?

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 15]
```diff
@@ -495,68 +482,81 @@
...
+                        aggregate-commit-index: none,
+                        votes-in-ustx: (get amount next-pool-signer),
+                        num-signer: u1,
                 })
-                
                 ;; Update pool map by appending wallet-candidate to list of candidates
                 (map-set pool next-cycle (merge 
                     next-pool
-                    {threshold-wallet-candidates: (unwrap! (as-max-len? (append (get threshold-wallet-candidates next-pool) pox-addr) u100) err-too-many-candidates)}
+                    {threshold-wallet-candidates: (unwrap! (as-max-len? (append (get threshold-wallet-candidates next-pool) pox-addr) u100) err-candidates-overflow)}
```
Comment (FriendsFerdinand) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:501]: There's nothing stopping a signer with a low stake in the protocol to fill the entire list with rubbish addresses.

Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:501]: In line 477 I assert that signer hasn't yet voted. 

So I believe that at most one signer with a low stake would be able to submit a single rubbish address? Or am I misunderstanding?

Either way we could / should check against the address by version I believe (though agreed I'll open another issue for that so this can get approved merged).

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 16]
```diff
@@ -495,68 +482,81 @@
...
+                                        (begin 
+                                            (print (/ (* new-candidate-votes u1000) next-pool-total-stacked))
+                                            false
+                                        )
+                                    )
+                                err-result
+                                    false
+                            )
+                        )
+                    ok-vote-existing-candidate-lost)
```
Comment (FriendsFerdinand) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:553]: This is the outcome of the branch where where someone voted correctly, no?
Why are we adding the `-lost` part here? That seems to indicate that you can't vote anymore after someone has lost.

Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:553]: Idea here is to keep branches returning parallel (ok ...) responses. 

Since a signer is voting for an existing candidate, that vote either pushes the existing candidate over the threshold-consensus, or it doesn't. 

If it does push existing candidate past consensus, we return (ok-vote-existing-candidate-won ...) - which is an ok vote for an existing candidate that won. 

If it doesn't push the existing candidate past consensus, we return (ok-vote-existing-candidate-lost ...) - which is an ok vote for an existing candidate that lost / didn't push the candidate past consensus.

Open to other naming conventions for these responses if you have any particular suggestion in mind.

Comment (FriendsFerdinand) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:553]: @setzeus Ok, it should be called `ok-vote-existing-candidate`

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 17]
```diff
@@ -727,14 +728,22 @@
...
+    (as-contract (fold mass-delegate-stack-extend (get stackers penalized-pool) (ok {stacker: tx-sender, unlock-burn-height: u0, pox-addr: pox-burn-address})))
+)
+
+;; mass delegate stack extender helper for either penalizing or concluded vote
+(define-private (mass-delegate-stack-extend (stacker principal) (pox-return (response {stacker: principal, unlock-burn-height: uint, pox-addr: { version: (buff 1), hashbytes: (buff 32) }} uint))) 
+    (let
+        (
+            (extend-delegate-response (unwrap! (contract-call? .pox-3 delegate-stack-extend stacker pox-burn-address u1) err-mass-delegate-stack-extend))
+            (extend-delegate-stacker (get stacker extend-delegate-response))
+            (extend-delegate-unlock-burn-height (get unlock-burn-height extend-delegate-response))
```
Comment (FriendsFerdinand) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:740]: Why are we always setting the `pox-addr` to the `pox-burn-address` ? Don't we have to use the `pox-addr` in the parameters?

File: sbtc-mini/contracts/sbtc-stacking-pool.clar [hunk 18]
```diff
@@ -727,14 +728,22 @@
...
+        )
+
+        ;; Assert previous pox-return was (ok ...)
+        (asserts! (is-ok pox-return) err-mass-delegate-stack-extend)
```
Comment (FriendsFerdinand) [inline sbtc-mini/contracts/sbtc-stacking-pool.clar:744]: We can `unwrap!` instead of doing 2 commands (`asserts!` and `is-ok`)

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 17 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar:740]: Yes. Great catch here, forget to update accordingly. Now outdated in latest commit:

https://github.com/Trust-Machines/stacks-sbtc/pull/578/commits/47e25a3065373ce482534db5410485e808892f93

File: sbtc-mini/contracts/sbtc-stacking-pool.clar (hunk 18 above)
Comment (setzeus) [reply sbtc-mini/contracts/sbtc-stacking-pool.clar]: Copy, fixed.

**Review by FriendsFerdinand — APPROVED** [review]

**Comment by MarvinJanssen** [comment]
Thanks @friedger
